cmake: Introduce `WITH_PYTHON` build option

hebasto opened the pull request:
This was suggested in https://github.com/bitcoin/bitcoin/issues/31476#issuecomment-2541288435:
> An alternative would be to require python to be disabled explicitly. Otherwise, it seems odd that every setting in cmake has a static default that can only be overridden explicitly, except for some, which are silently downgraded?

This PR updates the build system to fail by default on systems where the minimum required Python version is unavailable. It introduces an option that allows users to explicitly disable the Python requirement.

fanquake commented:
Not sure. Why does this need another build option? Given that this is just for working around silent CMake warnings in the CI, couldn't this just be fixed by using explicit `*_TEST` options that otherwise fail if Python is missing?

It also doesn't solve #31476 generically, because every time a similar case comes up, we'll have to add even more code and/or options.

hebasto commented:
Rebased due to the conflict with the merged #31651.

fanquake commented:
NACK - based on the reasoning above.

hebasto commented:
> NACK - based on the reasoning above.

> Not sure. Why does this need another build option? Given that this is just for working around silent CMake warnings in the CI, couldn't this just be fixed by using explicit `*_TEST` options that otherwise fail if Python is missing?

What "explicit `*_TEST`" do you mean?

fanquake commented:
> What "explicit *_TEST" do you mean?

(Introducing) An option that does anything test related, that also requires Python, but I don't think this solution is better than #31665, because it forces builders to start opting out of things, and doesn't solve #31476 generically, compared to #31665.

maflcko reviewed (approved):
lgtm. Seems fine to add an option to say "I don't have python and I don't care about the tests"

hebasto commented on CMakeLists.txt:
```diff
@@ -552,19 +552,20 @@ if(WERROR)
+  find_package(Python3 3.10 REQUIRED COMPONENTS Interpreter)
   set(PYTHON_COMMAND ${Python3_EXECUTABLE})
 else()
   list(APPEND configure_warnings
```
> shouldn't this be an error now?

Why? It precisely mirrors the scenario “I don't have python and I don't care about the tests” and simply serves as a reminder to the user of the consequences:

```
$ cmake -B build -DWITH_PYTHON=OFF
...
Use ccache for compiling .............. ON


  ******

CMake Warning at CMakeLists.txt:680 (message):
  Utils and rpcauth tests are disabled.

  Use "-DWITH_PYTHON=ON" to enable them.


  ******

-- Configuring done (12.8s)
-- Generating done (0.0s)
-- Build files have been written to: /home/hebasto/dev/bitcoin/build
```

An error happens when `WITH_PYTHON=ON` is set, which is the default, but Python has not been found.

hebasto replied:
@maflcko 
> Another alternative would be to just require python3, without a way to disable it. Thus, the option `WITH_PYTHON` can be removed.

Here is some historical context: treating the `Python3` package as optional was introduced in the staging [branch](https://github.com/hebasto/bitcoin/pull/15#discussion_r1192711084). In particular, @theuni noted:
> This makes us dependent on python, even if we're not running tests/bench, no? I wouldn't expect python to be a hard requirement for building libbitcoinkernel, for example.

That concern still appears to be relevant.

hebasto replied:
> I see. I guess I wanted to say the opposite: Can't this be removed, because it seems a bit odd to print a warning that something is disabled, when the users explicitly asked it to be disabled? There is also no warning when a user sets `BUILD_TESTS=OFF`.

Thanks! Implemented.

fanquake replied:
> That concern still appears to be relevant.

Yea. This is also essentially reintroducing the previous Autotools behaviour of having opt-out dependencies (i.e bdb for a long time), rather than opt-in, which is what we started doing with CMake.

hebasto committed:
cmake: Introduce `WITH_PYTHON` build option

By default, the build system generation will fail on systems where the
minimum required Python version is unavailable. This option allows
users to explicitly disable the Python requirement.

hebasto commented:
The [feedback](https://github.com/bitcoin/bitcoin/pull/31669#discussion_r2081903398) from @maflcko has been addressed.

Also rebased.

fanquake reviewed (changes requested):
As I mentioned above, adding more build options/compexity, and forcing builders to opt out of things they don't care about, does not seem like a good, nor generic approach to fixing #31476. Especially given this is essentially a CI fix, to work around the fact that CMake doesn't have a way to turn warnings into errors?

hebasto commented:
> > forcing builders to opt out of things they don't care about
> 
> I am not sure the current approach of silently skipping tests (with just a warning during configure, not when running the tests) is the right approach.
> 
> A third alternative would be to delay the error to when they run the tests.

FWIW, something similar has been implemented in https://github.com/bitcoin/bitcoin/pull/31233.

janb84 reviewed (commented):
Concept ACK 40d935904c2b1319b863306d32fef235cd008a8e
(if we want to keep python check in CMAKE)

PR Adds build option to ignore / not fail on missing python. When python is missing or below required version the configuration will error out. This removes failing silently on python.
This PR will partial solve issue #31476 (strictly the Python issue)  

---
Alternative thoughts: 
Should CMAKE even test for python ? If you see CMAKE as configuration for the BUILD step and we do not use python in building should we even test for python ?
